Fix Backend Compatibility

alyst opened the pull request:
Fixes #13 (matrix multiplication tests succeed on *ReverseDiff.jl*, but */* tests fail because it's not supported yet (IIUC))

Closes #15 (switches to `AbstractArray`)

Moves sparse matrix tests into extension (requires *MKLSparse.jl*).

Tests now include checks for type stability of the test functions.

alyst committed:
convert function lists to arrays

alyst committed:
[ul]triag_ldiv() require Julia>1.1

alyst committed:
bump version to 0.1.3

alyst committed:
fix type stability

alyst committed:
declare return type where inference fails

alyst committed:
relax arg types: Array -> AbstractArray

alyst committed:
replace matrix \ and / with LU-based functions

Julia doesn't allow arbitrary matrix to be used as-is in ldiv/rdiv

alyst committed:
move sparse_xxx_ldiv() funcs to extension

as it requires MKLSparse

alyst committed:
tests: return type inference

devmotion commented on src/DiffTests.jl:
```diff
@@ -25,15 +25,15 @@ names, one should not expect these functions to be "correct" for their original
 num2num_2(x) = 2*x + sqrt(x*x*x)
 num2num_3(x) = 10.31^(x + x) - x
-num2num_4(x) = 1
+num2num_4(x) = one(x)
```
Why is this needed?

alyst replied:
I assume that for num2num functions the contract is to return the same type as the input, so it could be tested in the unit tests along with type stability tests. It could be reverted, along with the tests.

alyst replied:
I reverted it

devmotion commented on src/DiffTests.jl:
```diff
@@ -25,15 +25,15 @@ names, one should not expect these functions to be "correct" for their original
 
-const NUMBER_TO_NUMBER_FUNCS = (num2num_1, num2num_2, num2num_3,
-                                num2num_4, num2num_5, identity)
+const NUMBER_TO_NUMBER_FUNCS = [num2num_1, num2num_2, num2num_3,
```
Can you explain why you switched the constant tuples to arrays?

alyst replied:
For consistency with matrix2matrix functions, which are extended in DiffTestsMKLSparseExt. I can revert num2num to be tuples again.

But out of curiosity - why they were tuples in the first place? What kind of optimisation does it trigger?

devmotion replied:
I assume because using tuples reduces allocations.

I don't think it matters much but I'd suggest reverting these changes in the PR because 1) they seem unrelated and 2) it is not clear that they improve anything (on the contrary, they increase allocations).

alyst replied:
I can revert it for the sets that are not modified in the extension. BTW, how much would it reduce memory overhead (considering that in return Julia would have to construct a tuple type)?

devmotion replied:
:shrug:

devmotion replied:
I'm a fan of keeping changes minimal and avoiding unrelated changes. So I'd suggest reverting these changes, regardless of whether they can/should be changed or not. For consistency, I think it would make sense to just use a tuple in the extension as well - and to open an issue or separate PR if you would like to change these tuples to arrays.

alyst replied:
It is possible to replace a tuple with the extended one (although I would then have to drop `const` from the declaration); the code would look a little bit more convoluted.
But if you mean that the extension should define its own set of sparse functions to test -- that would require adding support for that set of functions in all AD backend tests suites.

I am also a fan of minimal changes, but there are also consistency and maintainability considerations. For this PR I thought it would be strange to change some function sets into arrays and keep others as tuples, or rewrite tuples from the extension. But if you, as a maintainer, prefer minimal changes, I can minimize them.

devmotion replied:
Oh, I missed that the extension modifies arrays in the package. That does not seem OK to me, I don't think extensions are supposed to do that. For instance, this means that the resulting state depends on the order in which (possibly additional) extensions are loaded etc.

So in case you want to add an extension, I think a different design is needed anyway.

And yes, I very much would prefer more minimal changes.

alyst replied:
> That does not seem OK to me, I don't think extensions are supposed to do that. For instance, this means that the resulting state depends on the order in which (possibly additional) extensions are loaded etc.

What is the specific scenario that you think would be problematic?
The extensions would not remove functions from the lists, and they would be checking if the function is already in the list.
As for the order -- I don't think the order of functions in the list has any noticeable effects.
One thing to keep in mind is *Revise.jl*, but at the moment *DiffTests.jl* declares function lists as *const*, so *Revise.jl* cannot update current lists on the fly -- there is no regression here.

Also vector-based approach is used (see below) to dynamically populate the list based on Julia version (e.g. because some linear algebra routines are not available in the earlier versions). DiffTests.jl may drop support for Julia 1.0, but in the long run it might be beneficial to have some straightforward mechanism for supporting multiple Julia versions.

And let me refer to the original issue that I am trying to solve -- I've created JuliaDiff/ForwardDiff.jl#589 to address IMO the very relevant practical problem -- improving autodiff efficiency for linear models and multivariate normal distributions with dense or sparse matrices. So the test functions that I am adding here via the extension should be relevant for all autodiff backends. Vanilla *SparseArrays.jl* does not, at the moment, support *ldiv!()* for triangular sparse matrices, hence *MKLSparse.jl* weak dependency.

devmotion replied:
Extensions are just not designed for such a use case - their sole purpose is to extend existing functions.

A problematic scenario here would be another extension that also adds functions to the same array. Then the order of the functions in the array after loading the extensions can be completely arbitrary, and even worse if precompilation and loading of the extensions happens with multiple tasks in parallel two `push!`es might interfere.

devmotion commented on src/DiffTests.jl:
```diff
@@ -114,16 +114,16 @@ end
 self_weighted_logit(x) = inv(1.0 + exp(-dot(x, x)))
 
-nested_array_mul(x) = sum(sum(x[1] * [[x[2], x[3]]]))
+nested_array_mul(x)::eltype(x) = sum(sum(x[1] * [[x[2], x[3]]]))
```
The type annotation is wrong in general (one example is `x::Array{Bool}`).
```suggestion
nested_array_mul(x) = sum(sum(x[1] * [[x[2], x[3]]]))
```

alyst replied:
Yes, you are right, that's the shakiest part of the PR (this particular one could be fixed, but there is no single contract for the whole family of functions). My motivation was that if all functions are type stable, that might help with the testing times in ForwardDiff and especially ReverseDiff, which are abysmal.
Do you think it's a good thing to enforce type stability for these functions? How would you approach that?

devmotion replied:
They should be type stable if `sum` etc. are implemented in a type-stable way. But I guess maybe you rather mean that the type of the argument and the return type are the same/related?

devmotion replied:
Generally, I think the best is to remove all these return type annotations and first come up with concrete problems. Then it is also possible to figure out whether these problems should be fixed in DiffTests or e.g. in the AD packages.

alyst replied:
without this type annotation the `@inferred` macro, which I added to the tests, fails. IIUC for this function it is inferred to Any. I can drop these checks, or I can correct the type annotation, or adjust the function definition to be type stable. One real problem is the test times, so type stability may help a bit with that.

devmotion replied:
I'd suggest removing the `@inferred` statements and possibly opening an issue. I also don't know if there's much we can do here since these issues are caused by type inference problems in Base. On Julia 1.9.1 I get:
```julia
julia> using Test

julia> f(x) = x[1] * [[x[2], x[3]]]
f (generic function with 1 method)

julia> x = rand(3)
3-element Vector{Float64}:
 0.1648801068106326
 0.14897187763070263
 0.3251905884653816

julia> @inferred(f(x))
ERROR: return type Vector{Vector{Float64}} does not match inferred return type Any
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] top-level scope
   @ REPL[20]:1
```

alyst replied:
I have fixed the type inference of `nested_array_mul()`, it should be good for Bool-valued input vectors (however, only Float64 is tested at the moment, and some functions (mutating ones) are not compatible with Bool-valued vectors).

devmotion commented on src/DiffTests.jl:
```diff
@@ -247,49 +252,46 @@ function mutation_test_2!(y, x)
-sp_utriag_lmul(x::VecOrMat) = UpperTriangular(sparse(test_matrix(x))) * x
-sp_ltriag_lmul(x::VecOrMat) = LowerTriangular(sparse(test_matrix(x))) * x
+diag_lmul(x::AbstractVecOrMat) = diag_matrix(x) * x
+dense_lmul(x::AbstractVecOrMat)::typeof(x) = test_matrix(x) * x
```
Similarly, the return type here is not correct in general. Can it be removed?

alyst replied:
I've fixed these. The problem was that with `test_matrix = lehmer_matrix` type inference was broken; `test_matrix(x) = lehmer_matrix(x)` fixed that. So now all function definitions are type-stable, and we can keep `@inferred` in the tests.

devmotion commented on ext/DiffTestsMKLSparseExt.jl:
```diff
@@ -0,0 +1,25 @@
+module DiffTestsMKLSparseExt
+
+using LinearAlgebra: UpperTriangular, LowerTriangular
+using SparseArrays: sparse
```
BTW this should not be done: Extensions should only import the package itself and the weak dependency but not even standard libraries.

alyst replied:
Why? SparseArrays and LinearAlgebra are the direct dependencies of DiffTests itself and MKLSparse. The latter provides more specific definitions of matrix multiplication and division for sparse matrices that are utilized by the test functions.

devmotion replied:
You should import them indirectly via DiffTests and MKLSparse (see eg https://github.com/JuliaStats/LogExpFunctions.jl/pull/63).

alyst replied:
Thank you for the info. I've updated how the import is being done. However, the changes that the extensions makes to the function lists are not visible from the user session (I guess it's different world age of *DiffTests.jl* or something like that).

alyst committed:
fix type annotation

alyst committed:
revert num2num_4() change

alyst committed:
fix package import from ext

alyst committed:
fix type inference for matrix-based ops

alyst committed:
nested_array_mul(): improve type inference

alyst committed:
register() mechanism for function lists

const declaration of function lists does not work with extensions,
because these lists are redeclared each time the package module is
imported/used.
So these lists have to be defined inside __init__(), which is only
called once.

alyst committed:
sparse_ldiv(): disable

/ by generic sparse matrix is not supported

devmotion commented:
To fix https://github.com/JuliaDiff/DiffTests.jl/pull/17#discussion_r1238023180 properly, I had the following idea: Could we add traits to the test functions (eg., `istestfunc`, `isinplace`, `inputdim`, `outputdim` or whatever is appropriate) plus some convenience functions `number_to_number_funcs` etc. that basically just filter every function (or only inside of DiffTests?) according to the desired criteria? Then we could avoid pushing to fixed containers completely and would only extend the traits in the extension.

alyst commented:
> Could we add traits to the test functions (eg., istestfunc, isinplace, inputdim, outputdim or whatever is appropriate) plus some convenience functions number_to_number_funcs etc

I had a similar idea. But then, IIUC, to collect those functions one would have to use `methods()` -- this works fine, but the result is `MethodList`, and the elements are of type `Method`, not `Function`. `Method` is not callable, and at the first glance it looked like conversion would require some quirky low-level manipulations.

I figured out the problem I had with extension not updating the list of functions. It turned out that with each `import`/`using DiffTests` instruction, the `DiffTests` module was re-evaluated, i.e. all `const` assignments were reset. To fix that, I have moved assignments to `__init__()` (both in `DiffTests` and the extensions), which is called only once. Adding functions to the list is done with `DiffTests.register(registry_id, functions)` method, i.e. `DiffTests.register(:VECTOR_TO_VECTOR_FUNCS, [diag_lmul, utriag_lmul])`. The method checks that no duplicates are being introduced, and I can add critical section to ensure that parallel threads would not access it simultaneously. Under the hood, `register()` uses `setproperty!()` (wrapper for `setglobal!()` on 1.9, or `setfield!()` in earlier versions) to declare the corresponding global variable in `DiffTests` module, so no changes are required downstream.

alyst commented:
@devmotion I think the issue with updating the function lists by the extensions is solved. Do you have any further feedback on the PR? The functions functions that are being enabled are important for multivariate distributions.

devmotion commented:
Based on the previous comment, my impression is that there are still problems with e.g. simultaneous `register` calls. I'm generally skeptical regarding such global lists based on previous experiences with [e.g. DiffRules](https://github.com/JuliaDiff/DiffRules.jl/issues/90). Lastly, I'm a bit concerned adding a (weak) dependency that is broken on MacBooks (precompilation errors):
```julia
(jl_L4HrUa) pkg> precompile MKLSparse
Precompiling MKLSparse
  ✗ MKLSparse
  0 dependencies successfully precompiled in 1 seconds. 4 already precompiled.

ERROR: The following 1 direct dependency failed to precompile:

MKLSparse [0c723cd3-b8cd-5d40-b370-ba682dde9aae]

Failed to precompile MKLSparse [0c723cd3-b8cd-5d40-b370-ba682dde9aae] to "/Users/david/.julia/compiled/v1.9/MKLSparse/jl_aUHjJZ".
ERROR: LoadError: UndefVarError: `libmkl_rt` not defined
Stacktrace:
 [1] include(mod::Module, _path::String)
   @ Base ./Base.jl:457
 [2] include(x::String)
   @ MKLSparse ~/.julia/packages/MKLSparse/4LlPk/src/MKLSparse.jl:1
 [3] top-level scope
   @ ~/.julia/packages/MKLSparse/4LlPk/src/MKLSparse.jl:9
 [4] include
   @ ./Base.jl:457 [inlined]
 [5] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::Nothing)
   @ Base ./loading.jl:2049
 [6] top-level scope
   @ stdin:3
in expression starting at /Users/david/.julia/packages/MKLSparse/4LlPk/src/BLAS/BLAS.jl:1
in expression starting at /Users/david/.julia/packages/MKLSparse/4LlPk/src/MKLSparse.jl:1
in expression starting at stdin:3
```

alyst commented:
@devmotion Thanks for the feedback!

> I'm generally skeptical regarding such global lists based on previous experiences with https://github.com/JuliaDiff/DiffRules.jl/issues/90.

I agree that for highly extendable functionality, like *DiffRules.jl*, the current mechanism might be not flexible enough.
But our scope here is much more humble -- it is the package-internal mechanism to update lists of functions from its extensions (the `register()` method is not exported).
The implementation already updates the function lists from the extension's `__init__()` (the major limitation listed in the issue you referenced).
Basically, the only way the person can influence the lists of tested functions is via the project dependencies (which would activate the corresponding extensions).

> my impression is that there are still problems with e.g. simultaneous register calls.

I can add locks to ensure exclusive access if that is what you mean.

> Lastly, I'm a bit concerned adding a (weak) dependency that is broken on MacBooks (precompilation errors)

AFAIK there are issues with Mac version of *MKLSparse* with some of the linear algebra functions (see e.g. JuliaSparse/MKLSparse.jl#37), but the HEAD passes the tests.
The last release is 3 year-old, though (I am trying to fix that too). Which *MKLSparse.jl* version have you tried?
It's a bit hard to test across the architectures with GitHub CI right now, since it requires new releases of both *MKLSparse.jl* and *DiffTests.jl*.

devmotion commented:
MKL is not available for ARM, everyone with Apple silicon will run into the same issue (see, e.g., https://github.com/JuliaMath/FFTW.jl/issues/274).

devmotion commented:
Which makes me wonder: Maybe we should just not use MKL, and hence maybe we should just not add an extension. And thereby we could get rid of the design issues as well.

alyst commented:
> MKL is not available for ARM, everyone with Apple silicon will run into the same issue (see, e.g., https://github.com/JuliaMath/FFTW.jl/issues/274).
> Which makes me wonder: Maybe we should just not use MKL, and hence maybe we should just not add an extension. And thereby we could get rid of the design issues as well.

1. I would too prefer to avoid *MKLSparse.jl* dependency, but at the moment one runs into problems without it when dealing with certain models that have sparse covariation matrices.
2. I guess one of the goals of *JuliaDiff* ecosystem is to be extendable (or at least to support a wide range of applications). My experience is that right now it is not so easy.
   The PR is my attempt to improve it a little bit by providing a generic mechanism to update the list of tested functions based on the project dependencies.
    It is not flexible, but it is minimally disruptive to the current behaviour and package structure -- if one does not add these packages to the [test] project dependencies, nothing changes.
